numpyro/mcmc.py

```diff
@@ -428,3 +437,159 @@ def single_chain_mcmc(rng, init_params):
+
+    def sample(self, state):
+        return self.sample_fn(state)
+
+
+class NUTS(HMC):
+    def __init__(self,
+                 model=None,
+                 potential_fn=None,
+                 kinetic_fn=None,
+                 step_size=1.0,
+                 adapt_step_size=True,
+                 adapt_mass_matrix=True,
+                 dense_mass=False,
+                 target_accept_prob=0.8,
+                 trajectory_length=2 * math.pi,
+                 max_tree_depth=10):
+        super(NUTS, self).__init__(potential_fn=potential_fn, model=model, kinetic_fn=kinetic_fn,
+                                   step_size=step_size, adapt_step_size=adapt_step_size,
+                                   adapt_mass_matrix=adapt_mass_matrix, dense_mass=dense_mass,
+                                   target_accept_prob=target_accept_prob, trajectory_length=trajectory_length)
+        self.max_tree_depth = max_tree_depth
+        self.algo = 'NUTS'
+
+
+class MCMC(object):
+    def __init__(self,
+                 sampler,
+                 num_warmup,
+                 num_samples,
+                 num_chains=1,
+                 constrain_fn=None,
+                 progress_bar=True):
+        self.sampler = sampler
+        self.num_warmup = num_warmup
+        self.num_samples = num_samples
+        self.num_chains = num_chains
+        self.constrain_fn = constrain_fn
+        self.progress_bar = progress_bar
+        self.sequential_chain = False
+        if xla_bridge.device_count() < num_chains:
+            self.sequential_chain = True
+            warnings.warn('There are not enough devices to run parallel chains: expected {} but got {}.'
+                          ' Chains will be drawn sequentially. If you are running `mcmc` in CPU,'
+                          ' consider to disable XLA intra-op parallelism by setting the environment'
+                          ' flag "XLA_FLAGS=--xla_force_host_platform_device_count={}".'
+                          .format(num_chains, xla_bridge.device_count(), num_chains))
+        # TODO: We should have progress bars (maybe without diagnostics) for num_chains > 1
+        if num_chains > 1 or "CI" in os.environ or "PYTEST_XDIST_WORKER" in os.environ:
+            self.progress_bar = False
+
+        self._samples = None
+        self._samples_flat = None
+
+    def _single_chain_mcmc(self, rng, init_params=None, args=(), kwargs={}):
+        hmc_state, constrain_fn = self.sampler.init(rng, self.num_warmup, init_params,
+                                                    model_args=args, model_kwargs=kwargs)
+        if self.constrain_fn is None:
+            self.constrain_fn = identity if constrain_fn is None else constrain_fn
+        samples = fori_collect(self.num_warmup, self.num_warmup + self.num_samples,
+                               self.sampler.sample,
+                               hmc_state,
+                               transform=lambda x: self.constrain_fn(x.z),
```

Comment:
Collecting warmup samples should be a pretty straightforward addition. Reg. 

>Maybe we can have diagnostic_fn to kwargs of run method, that works like lambda x: (x.num_steps, x.step_size) and set transform=lambda x: (self.constrain_fn(x.z), self.diagnostic_fn())

Since we can collect all the HMC states (not just `z`), would it work if we specify the list of fields to collect through an arg like `collect_fields=('z', 'num_steps')`. We are storing `num_steps` in HMCState so that should be possible, but let me know if I am overlooking something.

Reply:
I would like to address both of these in a follow-up PR.

Reply:
> address both of these in a follow-up PR.
Definitely!

> collect through an arg like collect_fields=('z', 'num_steps')

This is much better than the thing I wanted, and easier to remember too. (using `transform=lambda x: x.adapt_state.step_size` is a bit unconventional, especially for users).